pkgs/development/go-modules/generic/default.nix

```diff
@@ -22,6 +22,7 @@
 , deleteVendor ? false
 # Whether to run the vend tool to regenerate the vendor directory.
 # This is useful if any dependency contain C files.
+# TODO: adapt to https://github.com/go-modules-by-example/index/blob/master/012_modvendor/README.md
```

Comment:
Wasn't this the previous approach that was replaced by the current implementation of `buildGoModule`?

Reply:
> Wasn't this the previous approach that was replaced by the current implementation of buildGoModule?

I don't know.

> where was the TODO discussed?

In my head and in https://github.com/NixOS/nixpkgs/pull/102426#issue-513794015 assuming that the upstream community is authoritative.

Reply:
There was [no previous approach](https://github.com/NixOS/nixpkgs/commit/95be7b25be151b6186eb1f8641458f175cc45e84).

Reply:
> There was no previous approach.

I'm referring to https://github.com/NixOS/nixpkgs/pull/93513

Reply:
The motivation of this PR really has it all. `TODO` might or might not be the right token for it. but `FUTURE` is not in any common _todo_ regex tooling, yet :wink:

Reply:
> I'm referring to #93513

Ah yeah that seems unrelated, though.

Reply:
We want to add `goDeps` / `vgo2nix` support for `buildGoModule` not another method using FOD.

Reply:
[`FOD`](https://en.wikipedia.org/wiki/Foreign_object_damage) ?

Reply:
> Ah yeah that seems unrelated, though.

No it isn't, it the PR that removed the method that you are now proposing here.

Reply:
Would that involve deprecating `vend`?

Reply:
fixed output derivation

Reply:
> No it isn't, it the PR that removed the method that you are now proposing here.

Reply:
How would https://github.com/nix-community/vgo2nix solve this?

Reply:
> I think you are mistaken

I'm not. See also https://github.com/NixOS/nixpkgs/pull/93513#discussion_r458510736

> How would https://github.com/nix-community/vgo2nix solve this?

We would use it and `goDeps` instead of vend.

Reply:
> See #93513 (comment)

Ah, I see where you are coming from, but in this comment, there seems to be a misunderstanding. The document proposes to use goproxy. The previous approach did not use goproxy.

Excerpt:

```
$ rm -rf modvendor
$ tgp=$(mktemp -d)
$ GOPROXY=file://$GOPATH/pkg/mod/cache/download GOPATH=$tgp go mod download
go: finding rsc.io/quote v1.5.2
go: finding rsc.io/sampler v1.3.0
go: finding golang.org/x/text v0.0.0-20170915032832-14c0d48ead0c
$ cp -rp $tgp/pkg/mod/cache/download/ modvendor
$ GOPATH=$tgp go clean -modcache
$ rm -rf $tgp
```
> Verify that modvendor can be used as a GOPROXY source:
```
GOPROXY=file:///home/gopher/scratchpad/modvendor_example/modvendor go run .
```

Reply:
> We would use it (cgo2nix)and `goDeps` instead of vend.

So this is superior to what the document proposes?

Reply:
> The previous approach did not use GOPROXY.

Yes, it did: https://github.com/NixOS/nixpkgs/commit/a0ddea1d6a9f07b0631e0dbcd3b7240d2eb82ef7#diff-e54d94124e13491694e0b3354c33d1b7d03f8576628447bc8140d05c862cbae3R115

Reply:
You mean this https://github.com/rogpeppe/godeps ?

Reply:
> You mean this https://github.com/rogpeppe/godeps ?

https://github.com/NixOS/nixpkgs/blob/5819bca301c78df2d05d69f76c259d41f74972e3/doc/languages-frameworks/go.xml#L113

Reply:
> Yes, it did: a0ddea1#diff-e54d94124e13491694e0b3354c33d1b7d03f8576628447bc8140d05c862cbae3R115

Ah ok, thank you! So I was mistaken. :wink: I might have got confused because little is known for the motivation. It says "dead code", that's all.

Reply:
So why is it deemed superior to ditch native tooling for custom one for dependency recording? To me it looks like a deliberate maintenance burden of the transliteration tooling. What am I missing?

Reply:
I'd suggest reading original PRs.

Reply:
I mean how can be a version controlled `go.mod` file not be trusted because of integrity concerns, for example?

Reply:
> I'd suggest reading original PRs.

There is no clear picture forming from [PR the history](https://github.com/NixOS/nixpkgs/commits/a90d267da81b9391cf3b36e1f4286461306846fc?before=a90d267da81b9391cf3b36e1f4286461306846fc+35&branch=a90d267da81b9391cf3b36e1f4286461306846fc&path%5B%5D=pkgs&path%5B%5D=development&path%5B%5D=go-modules&path%5B%5D=generic&path%5B%5D=default.nix).

From https://github.com/NixOS/nixpkgs/pull/16017, though:

> All program dependencies are extracted to json file than can include other json dependency files **for better package reuse** and **making overrides possible**.

https://github.com/NixOS/nixpkgs/pull/16017#issuecomment-229624046

> Because of the content-addressable store, if two programs have the same dependency it **will already result in the same derivation in the store**.

But dependencies are not built independently, so what do we gain? _Reuse_ probably not. So we are left with making overrides possible. Am I correct and complete?

Reply:
Maybe this, too? https://github.com/NixOS/nixpkgs/issues/100680

Reply:
https://discourse.nixos.org/t/creating-vendor-directories-directly-in-srcs-of-go-and-rust-packages-so-fixed-output-derivations-wont-be-needed/7367/5

Reply:
https://github.com/NixOS/nixpkgs/issues/52469#issuecomment-638772208 &rarr; seems like `goBuildModule` is the _blessed_ approach now a days: https://github.com/NixOS/nixpkgs/issues/52469#issuecomment-638772208

Reply:
So all in all, I'm concluding as that it appears that leaving this code comment is a) valid & b) useful and c) likely going in the direction the community will be taking, no?

Reply:
There is also: https://github.com/nlewo/nixpkgs/commit/09ac010e4fe4fb5cfa89a452912cc759dba20427#r43788902

And https://github.com/NixOS/nixpkgs/issues/84826#issuecomment-720286155

now.. :wink:

Reply:
> So all in all, I'm concluding as that it appears that leaving this code comment is a) valid & b) useful and c) likely going in the direction the community will be taking, no?

No, it is not. 

As I've already said, we want to add `goDeps / `vgo2nix` support to buildGoModule.

We don't want to re-adding this that (ab)uses FOD that we've already removed.

Reply:
Ok. I unserstand where one branch of the discussion is heading to. Thank you! 

Since it's clear folks will eventually change this, what alternative hint can we give, then at this line?

Please feel free to make a github suggestion to change this line to get this right.

Reply:
> We don't want to re-add this method that (ab)uses FOD.

In that regard you are comparing apples to oranges, don't you? This method is no different from the current implementation, the only difference is it also fetches non go files that reside in a directory that does not have go files.

But as an aside, any fetch is a fixed output derivation if the source and transport cannot be trusted.

So why the heck is it bad to have two fixed output derivations if building a go _module_ effectively involves two step fetching?

There is no chance one could catch the (real) problem of source immutability, anyways. For that nix community would have to host an instance of athens.